miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultWorksetService.java

```diff
@@ -0,0 +1,204 @@
+  }
+
+  public void setAuthorizationManager(AuthorizationManager authorizationManager) {
+    this.authorizationManager = authorizationManager;
+  }
+
+  @Override
+  public long count(Consumer<String> errorHandler, PaginationFilter... filter) throws IOException {
+    return worksetStore.count(errorHandler, filter);
+  }
+
+  @Override
+  public List<Workset> list(Consumer<String> errorHandler, int offset, int limit, boolean sortDir, String sortCol,
+      PaginationFilter... filter) throws IOException {
+    return worksetStore.list(errorHandler, offset, limit, sortDir, sortCol, filter);
+  }
+
+  @Override
+  public Workset get(long id) {
+    return worksetStore.get(id);
+  }
+
+  @Override
+  public List<Workset> listBySearch(String query) {
+    return worksetStore.listBySearch(query);
+  }
+
+  @Override
+  public List<Workset> listBySample(long sampleId) {
+    return worksetStore.listBySample(sampleId);
+  }
+
+  @Override
+  public List<Workset> listByLibrary(long libraryId) {
+    return worksetStore.listBySample(libraryId);
+  }
+
+  @Override
+  public List<Workset> listByDilution(long dilutionId) {
+    return worksetStore.listBySample(dilutionId);
+  }
+
+  @Override
+  public long save(Workset workset) throws IOException {
+    if (!workset.isSaved()) {
+      loadMembers(workset);
+      validateChange(workset, null);
+      workset.setChangeDetails(authorizationManager.getCurrentUser());
+      return worksetStore.save(workset);
+    } else {
+      Workset managed = worksetStore.get(workset.getId());
+      validateChange(workset, managed);
+      applyChanges(workset, managed);
+      managed.setChangeDetails(authorizationManager.getCurrentUser());
+      return worksetStore.save(managed);
+    }
+  }
+
+  private void loadMembers(Workset newWorkflow) throws IOException {
+    newWorkflow.setSamples(loadMembers("Sample", newWorkflow.getSamples(), WhineyFunction.rethrow(id -> sampleService.get(id))));
+    newWorkflow.setLibraries(loadMembers("Library", newWorkflow.getLibraries(), WhineyFunction.rethrow(id -> libraryService.get(id))));
+    newWorkflow.setDilutions(loadMembers("Dilution", newWorkflow.getDilutions(), WhineyFunction.rethrow(id -> dilutionService.get(id))));
+  }
+
+  private <T extends Identifiable> Set<T> loadMembers(String typeName, Collection<T> items, Function<Long, T> getter) {
+    Set<T> members = new HashSet<>();
+    for (T item : items) {
+      T member = getter.apply(item.getId());
+      if (member == null) {
+        throw new IllegalArgumentException(String.format("%s %d not found", typeName, item.getId()));
+      }
+      members.add(member);
+    }
+    return members;
+  }
+
+  private void applyChanges(Workset from, Workset to) throws IOException {
+    to.setAlias(from.getAlias());
+    to.setDescription(from.getDescription());
+    applyMemberChanges(from, to);
+  }
+
+  private void applyMemberChanges(Workset changed, Workset managed) throws IOException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.io.IOException', as it cannot be thrown from method's body. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)